### DISCUSSION TITLE
💥 1補2補大亂鬥!補數整理與分享

### DISCUSSION THREAD

**Discussion by JunTingLin** [1101-GS4719-C/Python-Community-GS4719-C#459] [General]
# 💥1補2補大亂鬥!補數整理與分享
## 1補2補運算
2的二進位為: 0010
|    原數        | 0010   |
| -------------- | -----  |
| 1補數(0、1互換) |  1101 |

|    原數        | 0010   |
| -------------- | -----  |
| 2補數(0、1互換+1) |  1110 |

## 11的1補(~11)是10進位的多少?

11的二進位為: 0000 1011
|    原數        | 0000 1011   |   說明          |
| -------------- | -----  |  ------     | 
| 1補數 | 1111 0100 |   這個數就是-12    |

看到第一碼是1，代表負值，後面其他的數必須用2補去轉，就變-12

## -11的1補(~-11)是10進位的多少?
因為是負的，所以要把11變2補

-11的二進位為: 1111 0101

|    原數        | 1111 0101   |   說明          |
| -------------- | -----  |  ------     | 
| 1補數 | 0000 1010 |   這個數就是10    |

## 1101它是負數嗎?我最左邊的bit要拿來當正負號嗎?

這個問題不是很明確，要看宣告是sign還是unsign

+ unsign
值的範圍0000(0)~1111(15)

+ sign
值的範圍1000(-8)~0111(7)

所以上面例子為例假如是11又須探討正負號，那四個bits無法表示，上面才會用8bits

## 電腦為何要2補數?
補數實際上是為了方便電腦在加減法處理上不需要額外的減法電路
例如:我們看到 `3 - 2 = 1`
但實際上電腦到的是 `3 + 2的補數`

0000 0011 + 1111 1110 = 1 0000 0001 (最前面的1 overflow掉了)
擠出去的1會被放在carry flag(CF)，我記的還有sign flag、zero flag、over flow flag，但我都還給老師了，再麻煩高人指點😁

**Comment by JunTingLin** [comment]
這邊補充一下CPU暫存器有許多flag，分想OF、CF
![image](https://user-images.githubusercontent.com/92431095/157601542-40aa808f-f4b3-4781-b8de-33e9e1213606.png)
以上圖為例53-45=8，在電腦裡運作要轉變成(HEX)35+D3= 108，最後的結果取8，1放在CF(carry flage)，OF是0(overflow flag)，

電腦靠這些就可以判斷它是合法的，值是+8😊

**Comment by JunTingLin** [comment]
電腦數學運算牽扯到幾個很重要的暫存器(EFLAGE)，其中有四個flag在組合語言數學運算式中都要追蹤，

CF（carry flag）:進位標誌 描述了最近操作是否發生了進位（可以檢查無符號操作是否溢位）unsign

ZF（zero flag）:零標誌 最近操作結果為0（列如 邏輯操作 等）

SF（sign flag）:符號標誌最近操作結果為負數

OF（overflow flag）:溢位標誌最近操作導致一個補碼溢位 補碼溢位通常有兩種結果（正溢位或者負溢位）sign

![image](https://user-images.githubusercontent.com/92431095/157603441-c2c64fea-fc68-48c7-903c-5f14c34bb230.png)

**Comment by 11timyyy** [comment]
謝謝，之前計概學的都差不多忘了XD!

**Comment by ghost** [comment]
讓我想起數導的老師有多爛QAQ
感謝分享
